### PR TITLE
Add multipolygon support to GeoPolygonQueryBuilder

### LINKED ISSUE #20789: Polygon query for geo_point field should pass the multi-polygon to LatLonPoint
This is a follow-on to #20315, where we cutover to `LatLonPoint` for faster geo shape filtering.

But we still make a big `BooleanQuery` when a multi-polygon (a union of top-level polygons, with or without internal holes) is passed.

This is in general far slower than passing the entire multi-polyon to `LatLonPoint`.

### PR DESCRIPTION
**Pull request by nknize**
LatLonPoint polygon query in Lucene 6 supports multipolygons and polygons with holes. This PR exposes this capability through the ES query DSL by adding `multipolygon` support to the `GeoPolygonQueryBuilder` parser such that the following can now be parsed:

```javascript
"query" : {
  "geo_polygon" : {
    "track.location" :{
      "multipolygon": [ [ [
        [-70, 40],
        [-80, 30],
        [-90, 20]
      ] ] ]
    }
  }
}
```
To maintain backcompat the following is still supported:

```javascript
"query" : {
  "geo_polygon" : {
    "track.location" :{
      "points": [ 
        [-70, 40],
        [-80, 30],
        [-90, 20]
      ] 
    }
  }
}
```

The `GeoPolygonQueryBuilder` now creates a `Polygon` object and passes it to `LatLonPoint.newPolygonQuery`. 

NOTE:

* Performs minimal polygon validation (e.g., enforcing minimum number of points, validating coordinates).
* Dateline crossing is not supported. 

TODO:

* update docs with multipolygon support

closes #20789

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/geo/GeoUtils.java [hunk 1]
```diff
@@ -454,6 +439,27 @@ public static GeoPoint parseGeoPoint(XContentParser parser, GeoPoint point) thro
...
 
+    public static GeoPoint parseNumericArray(XContentParser parser, GeoPoint point) throws IOException {
+        double lat = Double.NaN;
+        double lon = Double.NaN;
+        int element = 0;
+        while(parser.nextToken() != Token.END_ARRAY) {
+            if(parser.currentToken() == Token.VALUE_NUMBER) {
+                element++;
+                if(element == 1) {
+                    lon = parser.doubleValue();
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/common/geo/GeoUtils.java]: Hmm do we also do `lon` then `lat` elsewhere in the ES APIs?  This is very confusing ... Lucene standardized on `lat` then `lon` a while back!

File: core/src/main/java/org/elasticsearch/index/query/GeoPolygonQueryBuilder.java [hunk 2]
```diff
@@ -314,17 +390,26 @@ public static GeoPolygonQueryBuilder fromXContent(QueryParseContext parseContext
...
     public String getWriteableName() {
         return NAME;
     }
+
+    private void closeBoundary(List<GeoPoint> boundary) {
+        if (!boundary.get(boundary.size() - 1).equals(boundary.get(0))) {
+            boundary.add(boundary.get(0));
+        }
+        if (boundary.size() < 4) {
+            throw new IllegalArgumentException("too few points defined for geo_polygon query");
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/query/GeoPolygonQueryBuilder.java]: Can you include the actual points in the exc message to aid users debugging?

File: core/src/main/java/org/elasticsearch/index/query/GeoPolygonQueryBuilder.java [hunk 3]
```diff
@@ -168,53 +190,58 @@ protected Query doToQuery(QueryShardContext context) throws IOException {
...
-        // validation was not available prior to 2.x, so to support bwc
-        // percolation queries we only ignore_malformed on 2.x created indexes
-        if (!GeoValidationMethod.isIgnoreMalformed(validationMethod)) {
-            for (GeoPoint point : shell) {
-                if (!GeoUtils.isValidLatitude(point.lat())) {
-                    throw new QueryShardException(context, "illegal latitude value [{}] for [{}]", point.lat(),
-                            GeoPolygonQueryBuilder.NAME);
+        List<List<List<GeoPoint>>> polys = new ArrayList<>(this.polygons);
+        double[] shellLat = null, lat;
+        double[] shellLon = null, lon;
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/query/GeoPolygonQueryBuilder.java]: Hmm, `lat` and `lon` are not used in this outer scope, only down inside the nested `for` loop?  Can you move their declaration (the `double[]`) down there?

Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/query/GeoPolygonQueryBuilder.java]: Same for `shellLat` and `shellLon`?

File: core/src/main/java/org/elasticsearch/index/query/GeoPolygonQueryBuilder.java [hunk 4]
```diff
@@ -168,53 +190,58 @@ protected Query doToQuery(QueryShardContext context) throws IOException {
...
+        Polygon[] multiPoly = new Polygon[polys.size()];
+        int p = 0;
+        for (List<List<GeoPoint>> poly : polys) {
+            Polygon[] holes = new Polygon[poly.size() - 1];
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/query/GeoPolygonQueryBuilder.java]: Maybe assert `poly.size() > 0`?

File: core/src/main/java/org/elasticsearch/common/geo/GeoUtils.java [hunk 5]
```diff
@@ -454,6 +439,27 @@ public static GeoPoint parseGeoPoint(XContentParser parser, GeoPoint point) thro
...
+                } else if(element == 2) {
+                    lat = parser.doubleValue();
+                } else {
+                    throw new ElasticsearchParseException("only two values allowed");
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/common/geo/GeoUtils.java]: Can you add some context to this exception so the user knows what part of their request was wrong?

File: core/src/main/java/org/elasticsearch/common/geo/GeoUtils.java [hunk 6]
```diff
@@ -454,6 +439,27 @@ public static GeoPoint parseGeoPoint(XContentParser parser, GeoPoint point) thro
...
+                }
+            } else {
+                throw new ElasticsearchParseException("numeric value expected");
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/common/geo/GeoUtils.java]: Same here.

File: core/src/test/java/org/elasticsearch/index/query/GeoPolygonQueryBuilderTests.java [hunk 7]
```diff
@@ -230,6 +230,22 @@ public void testParsingAndToQuery4() throws IOException {
         assertGeoPolygonQuery(query);
     }
 
+    public void testParsingAndToQuery5() throws IOException {
```
Comment (mikemccand) [inline core/src/test/java/org/elasticsearch/index/query/GeoPolygonQueryBuilderTests.java]: It looks like these tests are a single polygon?  Can you also test that holes are parsed & executed correctly, and more than one polygon too?

File: core/src/main/java/org/elasticsearch/index/query/GeoPolygonQueryBuilder.java [hunk 8]
```diff
@@ -314,17 +390,26 @@ public static GeoPolygonQueryBuilder fromXContent(QueryParseContext parseContext
...
-        return Objects.hash(validationMethod, fieldName, shell, ignoreUnmapped);
+        return Objects.hash(validationMethod, fieldName, polygons, ignoreUnmapped);
     }
 
     @Override
     public String getWriteableName() {
         return NAME;
     }
+
+    private void closeBoundary(List<GeoPoint> boundary) {
```
Comment (mikemccand) [inline core/src/main/java/org/elasticsearch/index/query/GeoPolygonQueryBuilder.java]: Do we also close the boundary implicitly today as well?

**Review by mikemccand — COMMENTED** [review]
Thanks @nknize, I left a bunch of small comments.

It would be nice if ES would accept a standard GeoJson multipolygon: Lucene has a simple parser for this, taking `String` and returning `Polygon[]` ... I think that'd be easier for people to use than a triple-nested array of lon/lat points ... but we should do that separately.

**Comment by dakrone** [comment]
@nknize is this PR still a WIP? Is there someone else that should review it?

**Comment by islamjaidul** [comment]
I am sorry to say that I cannot use this way, it shows 
[parsing_exception] [geo_polygon] query does not support [multipolygon], with { line=1 & col=162 }

Elastic search version - 6.2.4
How may I use multipolygon, could you guys please tell me

**Comment by pcsanwald** [comment]
@polyfractal @nknize has this PR been supplanted by a different PR that provides the same functionality? if so, can you close please?

**Comment by polyfractal** [comment]
Quick update on this.  Nick rebased the PR onto master, looks like there are some CI failures related to backwards compatibility.  In general the PR is almost done but probably needs some more testing and better/more informative exception messages.

Nick doesn't have time to push this PR over the finish line right now, but we think it's still important so I'm going to mark this as a `help-me` issue.  Would make a good first issue for someone wanting to get involved with Geo, since most of the hard work is already done :)

**Comment by nknize** [comment]
This is a good first issue for anyone interested in picking up some ES geo knowledge. The only thing missing is to add support for GeoJSON using lucene's `Polygon.fromGeoJSON` method. Should be pretty straightforward.. just pass the GeoJSON string to that method and it will create the `Polygon[]` for you.

**Comment by ioanlucut** [comment]
Hello!

What is the status of this PR?

**Comment by nathancday** [comment]
If no one is working on this in 2020, I'm gonna pick it up and try to get it finalized

**Comment by nknize** [comment]
Thank you for the offer @nathancday but there is a 99% likelihood that we will be deprecating and removing `geo_polygon` query in favor of the added support for `geo_shape` queries on `geo_point` field types that was added in #53774 (which supports all geometry (and multipart) types). If that's the case then this PR is no longer needed and will be closed without merge.

**Comment by iverase** [comment]
I am going to close this as it does not apply any longer.

**Pull request closed without merging**